### ISSUE TITLE
Inappropriate type hint or obscured declaration error

### ISSUE THREAD

**Issue by Shapedsundew9** [microsoft/pylance-release#3834]
## Issue
Code snippet taken from [pylint example](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/dangerous-default-value.html)
The type hint given for penguin:
![image](https://user-images.githubusercontent.com/16618209/212375571-d368f297-137c-48d1-bbce-e6d3af9dec57.png)
if accepted, results in:
![image](https://user-images.githubusercontent.com/16618209/212375767-897a385f-160c-4f71-9e00-cab2649b0488.png)

## Environment data

Pylance language server 2023.1.20 (pyright b78a9e4b)

Ubuntu 22.04.1 LTS

VS Code
Version: 1.74.3
Commit: 97dec172d3256f8ca4bfb2143f3f76b503ca0534
Date: 2023-01-09T16:57:40.428Z
Electron: 19.1.8
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Linux x64 5.15.0-57-generic
Sandboxed: No

Python 3.11.1

##Code Snippet

```python
"""A doc string."""


def whats_on_the_telly(penguin: list[str] | None = None):
    """A doc string."""
    if penguin is None:
        penguin: list[str] = []
    penguin.append("property of the zoo")
    return penguin
```

## Expected behavior

Type definition not proposed or no obscured declaration error (not sure which is right).

## Actual behavior

Suggestion results in a problem for pylance.

## Logs
```
[Info  - 11:43:19 a.m.] (71551) Pylance language server 2023.1.20 (pyright b78a9e4b) starting
[Info  - 11:43:19 a.m.] (71551) Server root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.1.20/dist
[Info  - 11:43:19 a.m.] (71551) Starting service instance "scratch"
[Info  - 11:43:19 a.m.] (71551) No configuration file found.
[Info  - 11:43:19 a.m.] (71551) No pyproject.toml file found.
[Info  - 11:43:19 a.m.] (71551) Setting pythonPath for service "scratch": "/home/shapedsundew9/Projects/venv/bin/python3"
[Error - 11:43:19 a.m.] (71551) venvPath /home/shapedsundew9/Projects/scratch/~/Projects is not a valid directory.
[Warn  - 11:43:19 a.m.] (71551) stubPath /home/shapedsundew9/Projects/scratch/typings is not a valid directory.
[Info  - 11:43:19 a.m.] (71551) Assuming Python version 3.11
[Info  - 11:43:19 a.m.] (71551) Assuming Python platform Linux
[Info  - 11:43:19 a.m.] (71551) Searching for source files
[Info  - 11:43:19 a.m.] (71551) Found 1 source file
[Info  - 11:43:20 a.m.] (71551) Background analysis(1) root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.1.20/dist
[Info  - 11:43:20 a.m.] (71551) Background analysis(1) started
[Info  - 11:43:20 a.m.] (71551) Indexer background runner(2) root directory: /home/shapedsundew9/.vscode/extensions/ms-python.vscode-pylance-2023.1.20/dist (index)
[Info  - 11:43:20 a.m.] (71551) Indexing(2) started
[Info  - 11:43:21 a.m.] (71551) scanned(2) 256 files over 1 exec env
[Info  - 11:43:25 a.m.] (71551) [IDX(2)] Long operation: index execution environment /home/shapedsundew9/Projects/scratch (3447ms)
[Info  - 11:43:25 a.m.] (71551) [IDX(2)] Long operation: index packages /home/shapedsundew9/Projects/scratch (3459ms)
[Info  - 11:43:25 a.m.] (71551) indexed(2) 161 files over 1 exec env
[Info  - 11:43:25 a.m.] (71551) Indexing finished(2).
```

**Comment by erictraut** [comment]
In your sample, you have a symbol (a local variable) whose type is declared twice with incompatible type declarations. This isn't allowed. In one location, you're declaring that `penguin` can be assigned a value of type `list[str] | None`, but in another location, you're declaring that `penguin` can be assigned a value of type `list[str]`. It's the job of a type checker to enforce type declarations (i.e. verify that all assignments comply with the declared type). To do so, a type checker needs to know unambiguously which type you are declaring for a symbol. But in this case, you have declared two different types.

For more information about type declarations, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#type-declarations).

The easiest way to fix this bug is to delete the second (incompatible) type declaration for the symbol `penguin`. The type checker is smart enough to know that when you assign `[]` to the variable, its type is locally "narrowed" to just `list[str]`.

```python
def whats_on_the_telly(penguin: list[str] | None = None):
    """A doc string."""
    if penguin is None:
        penguin = []
    penguin.append("property of the zoo")
    return penguin
```

**Comment by Shapedsundew9** [comment]
That makes sense. So then I think the expected behaviour should be for pylance hinter not to propose that type and 'double click to add'.

**Comment by vineetrajasekhar** [comment]
Just curious - isn't this desirable/by design? 

It compels every appearance of a previously initialised name to comply with the first type annotation, and forbids the on-the-fly overwriting of a previously imposed type declaration with another... 

Or have I misunderstood?

**Comment by erictraut** [comment]
Yes, this is by design. You can choose to not declare the type of a symbol, in which case pyright (the type checker that underlies pylance) will allow any value to be assigned to it. But if you declare the type of a symbol, pyright will enforce that all assignments to that symbol comply with the type. After all, that's the job of a static type checker. You cannot declare a second (incompatible) type for a symbol because the type checker would no longer know which type to enforce for that symbol.

Static typing aside, it's generally not a good idea to use the same symbol name for values of different types anyway. It hurts code readability and maintainability.

For more information about type declarations, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md).

**Comment by debonte** [comment]
Just wanted to clarify that while the error is by design, Pylance shouldn't be showing an inlay hint for `penguin` on line 7. That's the bug.

**Comment by vineetrajasekhar** [comment]
Thank you both for explaining.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.3.22, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023322-15-March-2023-PreRelease)

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.3.22, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023322-15-March-2023-PreRelease)
